### Import Essential Libraries and dataset

In [1]:
#Basic Libraries For Data Extraction and Cleaning
import numpy as np
import pandas as pd

In [2]:
# Extract flight data from CSV
flightdata = pd.read_csv('Combined_Flights_2018.csv')

In [3]:
flightdata.head(30)

FlightDate            Airline Origin Dest  Cancelled  Diverted  \
0   2018-01-23  Endeavor Air Inc.    ABY  ATL      False     False   
1   2018-01-24  Endeavor Air Inc.    ABY  ATL      False     False   
2   2018-01-25  Endeavor Air Inc.    ABY  ATL      False     False   
3   2018-01-26  Endeavor Air Inc.    ABY  ATL      False     False   
4   2018-01-27  Endeavor Air Inc.    ABY  ATL      False     False   
5   2018-01-28  Endeavor Air Inc.    ABY  ATL      False     False   
6   2018-01-29  Endeavor Air Inc.    ABY  ATL      False     False   
7   2018-01-30  Endeavor Air Inc.    ABY  ATL      False     False   
8   2018-01-31  Endeavor Air Inc.    ABY  ATL      False     False   
9   2018-01-03  Endeavor Air Inc.    ATL  ABY      False     False   
10  2018-01-04  Endeavor Air Inc.    ATL  ABY      False     False   
11  2018-01-05  Endeavor Air Inc.    ATL  ABY      False     False   
12  2018-01-06  Endeavor Air Inc.    ATL  ABY      False     False   
13  2018-01-07  Endeavor Air Inc.    ATL  ABY      False     False   
14  2018-01-08  Endeavor Air Inc.    ATL  ABY      False     False   
15  2018-01-09  Endeavor Air Inc.    ATL  ABY      False     False   
16  2018-01-10  Endeavor Air Inc.    ATL  ABY      False     False   
17  2018-01-11  Endeavor Air Inc.    ATL  ABY      False     False   
18  2018-01-12  Endeavor Air Inc.    ATL  ABY      False     False   
19  2018-01-13  Endeavor Air Inc.    ATL  ABY      False     False   
20  2018-01-14  Endeavor Air Inc.    ATL  ABY      False     False   
21  2018-01-15  Endeavor Air Inc.    ATL  ABY      False     False   
22  2018-01-16  Endeavor Air Inc.    ATL  ABY      False     False   
23  2018-01-17  Endeavor Air Inc.    ATL  ABY       True     False   
24  2018-01-18  Endeavor Air Inc.    ATL  ABY      False     False   
25  2018-01-19  Endeavor Air Inc.    ATL  ABY      False     False   
26  2018-01-20  Endeavor Air Inc.    ATL  ABY      False     False   
27  2018-01-21  Endeavor Air Inc.    ATL  ABY      False     False   
28  2018-01-22  Endeavor Air Inc.    ATL  ABY      False     False   
29  2018-01-23  Endeavor Air Inc.    ATL  ABY      False     False   

    CRSDepTime  DepTime  DepDelayMinutes  DepDelay  ...  WheelsOff  WheelsOn  \
0         1202   1157.0              0.0      -5.0  ...     1211.0    1249.0   
1         1202   1157.0              0.0      -5.0  ...     1210.0    1246.0   
2         1202   1153.0              0.0      -9.0  ...     1211.0    1251.0   
3         1202   1150.0              0.0     -12.0  ...     1207.0    1242.0   
4         1400   1355.0              0.0      -5.0  ...     1412.0    1448.0   
5         1202   1202.0              NaN       NaN  ...     1236.0    1313.0   
6         1202   1204.0              2.0       2.0  ...     1219.0    1253.0   
7         1202   1153.0              0.0      -9.0  ...     1200.0    1244.0   
8         1202   1153.0              0.0      -9.0  ...     1219.0    1256.0   
9         1037   1101.0             24.0      24.0  ...     1124.0    1156.0   
10        1037   1032.0              0.0      -5.0  ...     1055.0    1122.0   
11        1037   1032.0              0.0      -5.0  ...     1052.0    1121.0   
12        1037   1032.0              0.0      -5.0  ...     1050.0    1124.0   
13        1037   1034.0              0.0      -3.0  ...     1052.0    1121.0   
14        1037   1059.0             22.0      22.0  ...     1117.0    1149.0   
15        1037   1222.0            105.0     105.0  ...     1242.0    1319.0   
16        1037   1027.0              0.0     -10.0  ...     1046.0    1123.0   
17        1037   1033.0              0.0      -4.0  ...     1104.0    1142.0   
18        1037   1044.0              7.0       7.0  ...     1100.0    1135.0   
19        1037   1034.0              0.0      -3.0  ...     1050.0    1124.0   
20        1037   1031.0              0.0      -6.0  ...     1047.0    1117.0   
21        1037   1030.0              0.0      -7.0  ...     1050.0    112

### Selection of Flight Status Dataset

We have selected the 2018 American flight log as it is during the pre-COVID-19 period and provides an accurate representation of flights data. From 2019 onwards, unpredictable and unprecedented factors due to the COVID-19 caused mass flight cancellations overnight.

### Identification of Response Variables

In this mini-project, we seek to establish and explore the potential contributing factors affecting the departure status (cancelled, on time, or delayed) of one's flight. Hence, the response variables that we have identified are the cancellation status and the extent to which a flight's departure is delayed. In order to achieve this, we will isolate the following variables as the response variables:

Cancelled [Boolean, Categorical]: True if flight is cancelled, False if flight is not cancelled

DepartureDelayGroup [Float, Categorical]: Represents departure delay time (Mins), in groups of 15 minutes. Negative value indicates early departure.

To smoothen the data, we will merge Cancelled with DepartureDelayGroup, so that the flight's status can be represented in just one response variable.

Because DepartureDelayGroup is of float data type, we cannot simply insert a boolean value to represent cancelled. Hence, we will take an arbitrary large value, 1000, to represent a cancelled flight. This is because DepartureDelayGroup does not contain a value as high as 1000 for delayed flights, as that would imply a delay of 15,000 minutes.  (Code further down in this notebook)

### Identification of Predictors

Secondly, we will select the potential variables that will be useful for the airlines customer to predict flight status:

Month [Int, Categorical]: Corresponds to the month of flight. (E.g. 1 == January)

DayOfWeek [Int, Categorical]: Corresponds to the day of the week (E.g. 2 == Tuesday)

Origin [Obj, Categorical]: Abbreviation of origin airport 

Dest [Obj, Categorical]: Abbreviation of destination airport

CRSDepTime [Int, Categorical]: Scheduled departure time

DistanceGroup [Int, Categorical]: Distance between origin and destination airport (Miles), in groups of 250 Miles

### Isolation of Predictors and Response Variables

In [4]:
new_flightdata = flightdata[['Month', 'DayOfWeek', 'Origin', 'Dest', 'CRSDepTime', 'DistanceGroup', 'Cancelled', 'DepartureDelayGroups']]
print("Number of rows:", new_flightdata.shape[0])
print("Number of columns:", new_flightdata.shape[1])

Number of rows: 5689512
Number of columns: 8


In [5]:
new_flightdata.head(10)

Month  DayOfWeek Origin Dest  CRSDepTime  DistanceGroup  Cancelled  \
0      1          2    ABY  ATL        1202              1      False   
1      1          3    ABY  ATL        1202              1      False   
2      1          4    ABY  ATL        1202              1      False   
3      1          5    ABY  ATL        1202              1      False   
4      1          6    ABY  ATL        1400              1      False   
5      1          7    ABY  ATL        1202              1      False   
6      1          1    ABY  ATL        1202              1      False   
7      1          2    ABY  ATL        1202              1      False   
8      1          3    ABY  ATL        1202              1      False   
9      1          3    ATL  ABY        1037              1      False   

   DepartureDelayGroups  
0                  -1.0  
1                  -1.0  
2                  -1.0  
3                  -1.0  
4                  -1.0  
5                   NaN  
6                   0.0  
7                  -1.0  
8                  -1.0  
9                   1.0

From the above data, we observed that for some rows, the data for DepartureDelayGroup is NaN even when the flight is not cancelled. This would mean that the actual departure time was not recorded, hence the departure delay was not calculated. 

Thus, we need to remove these rows to prevent any errors. In order to achieve this, we have created a helper function to remove any rows where the flight is not cancelled and predictor is NaN.

In [6]:
def removeNaNrows(column_name1,column_name2,df):
    # count the number of rows where column_name1 is NULL and column_name2 is False
    num_removed = len(df[(df[column_name1].isnull()) & (df[column_name2] == False)])
    df = df[~((df[column_name1].isnull()) & (df[column_name2] == False))]
    print("Removed " + str(num_removed) + " rows where " + column_name1 + " is NaN and " + column_name2 + " is False")
    return df

new_flightdata = flightdata[['Month', 'DayOfWeek', 'Origin', 'Dest', 'CRSDepTime', 'DistanceGroup', 'Cancelled', 'DepartureDelayGroups']]
new_flightdata = removeNaNrows('Month', 'Cancelled', new_flightdata)
new_flightdata = removeNaNrows('DayOfWeek', 'Cancelled',new_flightdata)
new_flightdata = removeNaNrows('Origin', 'Cancelled', new_flightdata)
new_flightdata = removeNaNrows('Dest', 'Cancelled', new_flightdata)
new_flightdata = removeNaNrows('CRSDepTime', 'Cancelled', new_flightdata)
new_flightdata = removeNaNrows('DistanceGroup', 'Cancelled', new_flightdata)
new_flightdata = removeNaNrows('Cancelled', 'Cancelled', new_flightdata)
new_flightdata = removeNaNrows('DepartureDelayGroups', 'Cancelled', new_flightdata)

Removed 0 rows where Month is NaN and Cancelled is False
Removed 0 rows where DayOfWeek is NaN and Cancelled is False
Removed 0 rows where Origin is NaN and Cancelled is False
Removed 0 rows where Dest is NaN and Cancelled is False
Removed 0 rows where CRSDepTime is NaN and Cancelled is False
Removed 0 rows where DistanceGroup is NaN and Cancelled is False
Removed 0 rows where Cancelled is NaN and Cancelled is False
Removed 1113 rows where DepartureDelayGroups is NaN and Cancelled is False


In [7]:
print("Number of rows after removal of anomalous rows:", new_flightdata.shape[0])
print("Number of columns:", new_flightdata.shape[1])

Number of rows after removal of anomalous rows: 5688399
Number of columns: 8


In [8]:
new_flightdata.head(10)

Month  DayOfWeek Origin Dest  CRSDepTime  DistanceGroup  Cancelled  \
0       1          2    ABY  ATL        1202              1      False   
1       1          3    ABY  ATL        1202              1      False   
2       1          4    ABY  ATL        1202              1      False   
3       1          5    ABY  ATL        1202              1      False   
4       1          6    ABY  ATL        1400              1      False   
6       1          1    ABY  ATL        1202              1      False   
7       1          2    ABY  ATL        1202              1      False   
8       1          3    ABY  ATL        1202              1      False   
9       1          3    ATL  ABY        1037              1      False   
10      1          4    ATL  ABY        1037              1      False   

    DepartureDelayGroups  
0                   -1.0  
1                   -1.0  
2                   -1.0  
3                   -1.0  
4                   -1.0  
6                    0.0  
7                   -1.0  
8                   -1.0  
9                    1.0  
10                  -1.0

### Integrating Cancellation Status into DepartureDelayGroup

As mentioned previously in this notebook, we will merge Cancelled into DepartureDelayGroups

In [9]:
# Set 'DepartureDelayGroups' to 1000 where 'cancelled' is True
new_flightdata.loc[new_flightdata['Cancelled'], 'DepartureDelayGroups'] = 1000
new_flightdata = new_flightdata.drop('Cancelled', axis=1)

DepartureDelayGroup [UPDATED] [Float, Categorical]: Represents departure delay time (Mins), in groups of 15 minutes. Negative value indicates early departure. +1000 indicates cancelled flight.

In [10]:
new_flightdata.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5688399 entries, 0 to 5689511
Data columns (total 7 columns):
 #   Column                Dtype  
---  ------                -----  
 0   Month                 int64  
 1   DayOfWeek             int64  
 2   Origin                object 
 3   Dest                  object 
 4   CRSDepTime            int64  
 5   DistanceGroup         int64  
 6   DepartureDelayGroups  float64
dtypes: float64(1), int64(4), object(2)
memory usage: 347.2+ MB


Our cleaned Dataset consists of 6 variables, with a total of 5688399 rows. 

In [11]:
# Convert DataFrame to CSV file
new_flightdata.to_csv('cleaned_data.csv')